pyquil/quilatom.py

```diff
@@ -899,26 +900,58 @@ def out(self) -> str:
         return str(self)
 
 
-@dataclass
-class WaveformReference(QuilAtom):
+class WaveformReference(quil_rs.WaveformInvocation):
```

Comment:
Out of scope for this PR but relevant: my feeling is that all waveform invocations should be opaque waveform references with `Dict[string, Expression]` parameters, because validating these on the client can leave the client and service out of sync, and the service's opinion is the one that matters.

Say we add support for a new `radical` waveform in compilation, and it takes parameter `stoke` but no `duration`. Anyone with this version of pyQuil or earlier just wouldn't be able to use it, right? Even though `PULSE ... radical(stoke: 5000)` is valid Quil and would compile correctly.

Reply:
Yeah, I agree,`quil-rs` does the right thing with `WaveformInvocation` and it's unfortunate pyQuil doesn't have an equivalent. The way I reworked `TemplateWaveform` almost gets us there in that it can now take any set of parameters, but as you say, `duration` is required since it was part of the API for that class. Since we are deprecating the `TemplateWaveform` stuff, I'll throw in a thin wrapper for `WaveformInvocation` and point users to it in the deprecation warning.